File: lection03/app_java/src/main/java/ru/hse/lection03/businesslayer/DroidRepository.java
package ru.hse.lection03.businesslayer;

import java.util.ArrayList;
import java.util.List;

public class DroidRepository {
    public static Short DATA_SIZE ;

    private static volatile DroidRepository mInstance;

    public static DroidRepository getInstance() {
        if (mInstance == null) {
            synchronized (DroidRepository.class) {
                if (mInstance == null) {
                    mInstance = new DroidRepository();
                }
            }
        }
        return mInstance;
    }


    protected final List<Short> mData;


    private DroidRepository() {
        mData = initializeData();
    }


    // получить список Дроидов
    public List<Short> list() {
        return mData;
    }

    // получить дроида по индексу
    public Short item(int index) {
        return mData.get(index);
    }

    public void addNewNumber() {
        mData.add(++DATA_SIZE);
    }

    // Функция инициализации списка дроидов
    protected List<Short> initializeData() {
        final List<Short> data = new ArrayList<>();

        // Наполняем лист в цикле
        for (short position = 0; position <= 100; position ++) {
            data.add(position);
        }
        DATA_SIZE=100;
        return data;
    }
}

File: lection03/app_kotlin/src/main/java/ru/hse/lection03/businesslayer/DroidRepository.kt
package ru.hse.lection03.businesslayer

import ru.hse.lection03.objects.Droid
import java.util.*

class DroidRepository private constructor() {
    companion object {
        // Константы
        const val DATA_SIZE = 100
        const val STATE_CRITICAL = 0.8f

        // простенький singleton
        val instance by lazy { DroidRepository() }

        // Для генерации случайных чисел
        protected val RANOMIZER = Random()
    }


    // ленивая инициализации для списка Дроидов
    protected val droidList by lazy { initializeData() }


    // получить список Дроидов
    fun list() = droidList

    // получить дроида по индексу
    fun item(index: Int) = droidList[index]


    // Функция инициализации списка дроидов
    protected fun initializeData(): List<Droid> {
        val data = mutableListOf<Droid>()

        // Наполняем лист в цикле
        for (position in 0 until DATA_SIZE) {

            // Генерим имя дроида
            val name = "Droid $position"

            // Получаем случайное число, и определяем состояние дроида
            val state = when (RANOMIZER.nextFloat() >= STATE_CRITICAL) {
                true -> Droid.STATE_REMOVED
                false -> Droid.STATE_NEW
            }

            // Создаем дроида и добавляем его в список
            val droid = Droid(name, state)
            data.add(droid)
        }

        return data
    }
}